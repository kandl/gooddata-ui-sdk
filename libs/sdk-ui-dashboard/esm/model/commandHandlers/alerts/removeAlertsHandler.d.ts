import { SagaIterator } from "redux-saga";
import { DashboardContext } from "../../types/commonTypes.js";
import { RemoveAlerts } from "../../commands/alerts.js";
import { DashboardAlertsRemoved } from "../../events/alerts.js";
export declare function removeAlertsHandler(ctx: DashboardContext, cmd: RemoveAlerts): SagaIterator<DashboardAlertsRemoved>;
