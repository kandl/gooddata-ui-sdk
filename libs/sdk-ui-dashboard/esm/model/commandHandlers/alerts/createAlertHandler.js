import { call, put } from "redux-saga/effects";
import { alertCreated } from "../../events/alerts.js";
import { alertsActions } from "../../store/alerts/index.js";
function createAlert(ctx, alert) {
    const { backend, workspace } = ctx;
    return backend.workspace(workspace).dashboards().createWidgetAlert(alert);
}
export function* createAlertHandler(ctx, cmd) {
    const alert = yield call(createAlert, ctx, cmd.payload.alert);
    yield put(alertsActions.addAlert(alert));
    return alertCreated(ctx, alert, cmd.correlationId);
}
//# sourceMappingURL=createAlertHandler.js.map