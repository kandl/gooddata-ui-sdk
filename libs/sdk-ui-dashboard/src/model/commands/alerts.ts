// (C) 2021-2023 GoodData Corporation

import { ObjRef, IWidgetAlert, IWidgetAlertDefinition } from "@gooddata/sdk-model";
import { IDashboardCommand } from "./base.js";

/**
 * Payload of the {@link CreateAlert} command.
 * @beta
 */
export interface CreateAlertPayload {
    /**
     * The alert to be created.
     */
    readonly alert: IWidgetAlertDefinition;
}

/**
 * Creates Kpi alert.
 *
 * @beta
 */
export interface CreateAlert extends IDashboardCommand {
    readonly type: "GDC.DASH/CMD.ALERT.CREATE";
    readonly payload: CreateAlertPayload;
}

/**
 * Creates the SaveAlert command. Dispatching this command will result in the creating Kpi alert on the backend.
 *
 * @param alert - specify alert to create.
 * @param correlationId - specify correlation id to use for this command. this will be included in all
 *  events that will be emitted during the command processing
 * @beta
 */
export function createAlert(alert: IWidgetAlertDefinition, correlationId?: string): CreateAlert {
    return {
        type: "GDC.DASH/CMD.ALERT.CREATE",
        correlationId,
        payload: {
            alert,
        },
    };
}

//
//
//

/**
 * Payload of the {@link UpdateAlert} command.
 * @beta
 */
export interface UpdateAlertPayload {
    /**
     * The alert to be updated.
     */
    readonly alert: IWidgetAlert;
}

/**
 * Updates Kpi alert.
 *
 * @beta
 */
export interface UpdateAlert extends IDashboardCommand {
    readonly type: "GDC.DASH/CMD.ALERT.UPDATE";
    readonly payload: UpdateAlertPayload;
}

/**
 * Creates the UpdateAlert command. Dispatching this command will result in the updating Kpi alert on the backend.
 *
 * @param alert - specify alert to update.
 * @param correlationId - specify correlation id to use for this command. this will be included in all
 *  events that will be emitted during the command processing

 * @beta
 */
export function updateAlert(alert: IWidgetAlert, correlationId?: string): UpdateAlert {
    return {
        type: "GDC.DASH/CMD.ALERT.UPDATE",
        correlationId,
        payload: {
            alert,
        },
    };
}

//
//
//

/**
 * Payload of the {@link RemoveAlerts} command.
 * @beta
 */
export interface RemoveAlertsPayload {
    /**
     * References to Kpi alerts that should be removed.
     */
    readonly refs: ObjRef[];
}

/**
 * Removes Kpi alerts.
 *
 * @beta
 */
export interface RemoveAlerts extends IDashboardCommand {
    readonly type: "GDC.DASH/CMD.ALERTS.REMOVE";
    readonly payload: RemoveAlertsPayload;
}

/**
 * Creates the RemoveAlerts command. Dispatching this command will result in the removing Kpi alerts on the backend.
 *
 * @param refs - specify ObjRef of the alerts to remove
 * @param correlationId - specify correlation id to use for this command. this will be included in all
 *  events that will be emitted during the command processing

 * @beta
 */
export function removeAlerts(refs: ObjRef[], correlationId?: string): RemoveAlerts {
    return {
        type: "GDC.DASH/CMD.ALERTS.REMOVE",
        correlationId,
        payload: {
            refs,
        },
    };
}
