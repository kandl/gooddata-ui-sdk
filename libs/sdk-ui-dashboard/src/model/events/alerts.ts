// (C) 2021-2023 GoodData Corporation
import { IWidgetAlert } from "@gooddata/sdk-model";

import { DashboardContext } from "../types/commonTypes.js";
import { IDashboardEvent } from "./base.js";
import { eventGuard } from "./util.js";

/**
 * Payload of the {@link DashboardAlertCreated} event.
 * @beta
 */
export interface DashboardAlertCreatedPayload {
    /**
     * The alert created.
     */
    readonly alert: IWidgetAlert;
}

/**
 * This event is emitted after the Kpi alert is successfully saved.
 *
 * @beta
 */
export interface DashboardAlertCreated extends IDashboardEvent {
    readonly type: "GDC.DASH/EVT.ALERT.CREATED";
    readonly payload: DashboardAlertCreatedPayload;
}

export function alertCreated(
    ctx: DashboardContext,
    alert: IWidgetAlert,
    correlationId?: string,
): DashboardAlertCreated {
    return {
        type: "GDC.DASH/EVT.ALERT.CREATED",
        ctx,
        correlationId,
        payload: {
            alert,
        },
    };
}

/**
 * Tests whether the provided object is an instance of {@link DashboardAlertCreated}.
 *
 * @param obj - object to test
 * @beta
 */
export const isDashboardAlertCreated = eventGuard<DashboardAlertCreated>("GDC.DASH/EVT.ALERT.CREATED");

/**
 * Payload of the {@link DashboardAlertsRemoved} event.
 * @beta
 */
export interface DashboardAlertsRemovedPayload {
    /**
     * The alerts removed.
     */
    readonly alerts: IWidgetAlert[];
}

/**
 * This event is emitted after the Kpi alerts are successfully removed.
 *
 * @beta
 */
export interface DashboardAlertsRemoved extends IDashboardEvent {
    readonly type: "GDC.DASH/EVT.ALERTS.REMOVED";
    readonly payload: DashboardAlertsRemovedPayload;
}

export function alertsRemoved(
    ctx: DashboardContext,
    alerts: IWidgetAlert[],
    correlationId?: string,
): DashboardAlertsRemoved {
    return {
        type: "GDC.DASH/EVT.ALERTS.REMOVED",
        ctx,
        correlationId,
        payload: {
            alerts,
        },
    };
}

/**
 * Tests whether the provided object is an instance of {@link DashboardAlertsRemoved}.
 *
 * @param obj - object to test
 * @beta
 */
export const isDashboardAlertsRemoved = eventGuard<DashboardAlertsRemoved>("GDC.DASH/EVT.ALERTS.REMOVED");

/**
 * Payload of the {@link DashboardAlertUpdated} event.
 * @beta
 */
export interface DashboardAlertUpdatedPayload {
    /**
     * The alert updated.
     */
    readonly updated: IWidgetAlert;
}

/**
 * This event is emitted after the Kpi alert is updated.
 *
 * @beta
 */
export interface DashboardAlertUpdated extends IDashboardEvent {
    readonly type: "GDC.DASH/EVT.ALERT.UPDATED";
    readonly payload: DashboardAlertUpdatedPayload;
}

export function alertUpdated(
    ctx: DashboardContext,
    updated: IWidgetAlert,
    correlationId?: string,
): DashboardAlertUpdated {
    return {
        type: "GDC.DASH/EVT.ALERT.UPDATED",
        ctx,
        correlationId,
        payload: {
            updated,
        },
    };
}

/**
 * Tests whether the provided object is an instance of {@link DashboardAlertUpdated}.
 *
 * @param obj - object to test
 * @beta
 */
export const isDashboardAlertUpdated = eventGuard<DashboardAlertUpdated>("GDC.DASH/EVT.ALERT.UPDATED");
