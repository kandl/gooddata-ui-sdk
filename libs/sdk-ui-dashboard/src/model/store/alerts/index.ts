// (C) 2021 GoodData Corporation

import { createSlice } from "@reduxjs/toolkit";
import { alertsAdapter } from "./alertsEntityAdapter.js";

const alertsSlice = createSlice({
    name: "alerts",
    initialState: alertsAdapter.getInitialState(),
    reducers: {
        setAlerts: alertsAdapter.setAll,
        addAlert: alertsAdapter.addOne,
        updateAlert: alertsAdapter.updateOne,
        removeAlert: alertsAdapter.removeOne,
        removeAlerts: alertsAdapter.removeMany,
    },
});

export const alertsSliceReducer = alertsSlice.reducer;
export const alertsActions = alertsSlice.actions;
