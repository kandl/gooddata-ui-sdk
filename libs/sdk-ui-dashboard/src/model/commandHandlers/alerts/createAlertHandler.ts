// (C) 2021-2022 GoodData Corporation
import { SagaIterator } from "redux-saga";
import { call, put } from "redux-saga/effects";
import { DashboardContext } from "../../types/commonTypes.js";
import { IWidgetAlert, IWidgetAlertDefinition } from "@gooddata/sdk-model";
import { CreateAlert } from "../../commands/alerts.js";
import { alertCreated, DashboardAlertCreated } from "../../events/alerts.js";
import { PromiseFnReturnType } from "../../types/sagas.js";
import { alertsActions } from "../../store/alerts/index.js";

function createAlert(
    ctx: DashboardContext,
    alert: IWidgetAlert | IWidgetAlertDefinition,
): Promise<IWidgetAlert> {
    const { backend, workspace } = ctx;

    return backend.workspace(workspace).dashboards().createWidgetAlert(alert);
}

export function* createAlertHandler(
    ctx: DashboardContext,
    cmd: CreateAlert,
): SagaIterator<DashboardAlertCreated> {
    const alert: PromiseFnReturnType<typeof createAlert> = yield call(createAlert, ctx, cmd.payload.alert);

    yield put(alertsActions.addAlert(alert));

    return alertCreated(ctx, alert, cmd.correlationId);
}
