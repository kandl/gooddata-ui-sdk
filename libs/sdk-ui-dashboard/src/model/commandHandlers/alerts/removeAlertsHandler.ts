// (C) 2021 GoodData Corporation
import { SagaIterator } from "redux-saga";
import { call, select, put } from "redux-saga/effects";
import { ObjRef, objRefToString } from "@gooddata/sdk-model";
import { DashboardContext } from "../../types/commonTypes.js";
import { RemoveAlerts } from "../../commands/alerts.js";
import { alertsRemoved, DashboardAlertsRemoved } from "../../events/alerts.js";
import { alertsActions } from "../../store/alerts/index.js";
import { selectAlertsMap } from "../../store/alerts/alertsSelectors.js";
import { validateExistingAlerts } from "./validation/alertsValidation.js";

function removeAlerts(ctx: DashboardContext, alertRefs: ObjRef[]): Promise<void> {
    const { backend, workspace } = ctx;
    return backend.workspace(workspace).dashboards().deleteWidgetAlerts(alertRefs);
}

export function* removeAlertsHandler(
    ctx: DashboardContext,
    cmd: RemoveAlerts,
): SagaIterator<DashboardAlertsRemoved> {
    const alertsMap: ReturnType<typeof selectAlertsMap> = yield select(selectAlertsMap);
    const alerts = validateExistingAlerts(alertsMap, cmd, ctx);

    const refs = alerts.map((alert) => alert.ref);

    yield call(removeAlerts, ctx, refs);

    const ids = alerts.map((alert) => objRefToString(alert.ref));

    yield put(alertsActions.removeAlerts(ids));

    return alertsRemoved(ctx, alerts, cmd.correlationId);
}
