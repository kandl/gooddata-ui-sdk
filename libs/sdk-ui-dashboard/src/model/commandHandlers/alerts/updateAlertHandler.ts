// (C) 2021-2022 GoodData Corporation
import { SagaIterator } from "redux-saga";
import { call, put } from "redux-saga/effects";
import { objRefToString, IWidgetAlert } from "@gooddata/sdk-model";
import { DashboardContext } from "../../types/commonTypes.js";
import { UpdateAlert } from "../../commands/alerts.js";
import { alertUpdated, DashboardAlertUpdated } from "../../events/alerts.js";
import { PromiseFnReturnType } from "../../types/sagas.js";
import { alertsActions } from "../../store/alerts/index.js";

function updateAlert(ctx: DashboardContext, alert: IWidgetAlert): Promise<IWidgetAlert> {
    const { backend, workspace } = ctx;

    return backend.workspace(workspace).dashboards().updateWidgetAlert(alert);
}

export function* updateAlertHandler(
    ctx: DashboardContext,
    cmd: UpdateAlert,
): SagaIterator<DashboardAlertUpdated> {
    const alert: PromiseFnReturnType<typeof updateAlert> = yield call(updateAlert, ctx, cmd.payload.alert);

    yield put(
        alertsActions.updateAlert({
            changes: cmd.payload.alert,
            id: objRefToString(cmd.payload.alert.ref),
        }),
    );

    return alertUpdated(ctx, alert, cmd.correlationId);
}
