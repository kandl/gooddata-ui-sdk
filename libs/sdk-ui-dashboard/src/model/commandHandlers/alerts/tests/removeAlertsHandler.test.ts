// (C) 2021-2022 GoodData Corporation

import { DashboardTester, preloadedTesterFactory } from "../../../tests/DashboardTester.js";
import { idRef, uriRef } from "@gooddata/sdk-model";
import { RemoveAlerts, removeAlerts } from "../../../commands/alerts.js";
import {
    BrokenFilterAlertsDashboardIdentifier,
    AlertForRemovedFiltersKpi,
    TestCorrelation,
} from "../../../tests/fixtures/BrokenFilterAlerts.fixtures.js";
import { DashboardAlertsRemoved } from "../../../events/alerts.js";
import { selectAlertByRef } from "../../../store/alerts/alertsSelectors.js";
import { DashboardCommandFailed } from "../../../events/index.js";
import { beforeEach, describe, it, expect } from "vitest";

describe("removeDrillsForInsightWidgetHandler", () => {
    const alertRef = AlertForRemovedFiltersKpi.ref!;
    const alertsRefsArray = [alertRef];

    let Tester: DashboardTester;

    beforeEach(async () => {
        await preloadedTesterFactory((tester) => {
            Tester = tester;
        }, BrokenFilterAlertsDashboardIdentifier);
    });

    describe("remove", () => {
        it("should emit the appropriate events for remove alerts and propagate correlationId", async () => {
            await Tester.dispatchAndWaitFor(
                removeAlerts(alertsRefsArray, TestCorrelation),
                "GDC.DASH/EVT.ALERTS.REMOVED",
            );
            expect(Tester.emittedEventsDigest()).toMatchSnapshot();
        });

        it("should remove alerts from the state and propagate it via event payload", async () => {
            const alertSelector = selectAlertByRef(alertRef);

            const alertFromStateBefore = alertSelector(Tester.state());
            expect(alertFromStateBefore).toBeTruthy();

            const event: DashboardAlertsRemoved = await Tester.dispatchAndWaitFor(
                removeAlerts(alertsRefsArray, TestCorrelation),
                "GDC.DASH/EVT.ALERTS.REMOVED",
            );

            expect(event.payload.alerts).toEqual([AlertForRemovedFiltersKpi]);

            const alertFromState = alertSelector(Tester.state());
            expect(alertFromState).toBeFalsy();
        });

        it("should remove alerts referenced by Identifer from the state and propagate it via event payload", async () => {
            const alertIdentifier = idRef(AlertForRemovedFiltersKpi.identifier!);

            const alertSelector = selectAlertByRef(alertRef);

            const alertFromStateBefore = alertSelector(Tester.state());
            expect(alertFromStateBefore).toBeTruthy();

            const event: DashboardAlertsRemoved = await Tester.dispatchAndWaitFor(
                removeAlerts([alertIdentifier], TestCorrelation),
                "GDC.DASH/EVT.ALERTS.REMOVED",
            );

            expect(event.payload.alerts).toEqual([AlertForRemovedFiltersKpi]);

            const alertFromState = alertSelector(Tester.state());
            expect(alertFromState).toBeFalsy();
        });
    });

    describe("validate", () => {
        it("should fail if trying to remove non-existent alert", async () => {
            const event: DashboardCommandFailed<RemoveAlerts> = await Tester.dispatchAndWaitFor(
                removeAlerts([uriRef("missing")], TestCorrelation),
                "GDC.DASH/EVT.COMMAND.FAILED",
            );

            expect(event.payload.reason).toEqual("USER_ERROR");
            expect(event.correlationId).toEqual(TestCorrelation);
        });
    });
});
