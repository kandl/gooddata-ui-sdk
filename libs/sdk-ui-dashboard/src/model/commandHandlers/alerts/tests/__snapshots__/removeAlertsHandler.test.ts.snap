// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`removeDrillsForInsightWidgetHandler > remove > should emit the appropriate events for remove alerts and propagate correlationId 1`] = `
[
  {
    "commandType": "GDC.DASH/CMD.ALERTS.REMOVE",
    "correlationId": "testCorrelationId",
    "type": "GDC.DASH/EVT.COMMAND.STARTED",
  },
  {
    "correlationId": "testCorrelationId",
    "type": "GDC.DASH/EVT.ALERTS.REMOVED",
  },
]
`;
